rama-net/src/fingerprint/peet/tls.rs

```diff
@@ -0,0 +1,625 @@
+    pub fn compute_from_client_hello(
+        client_hello: impl ClientHelloProvider,
+    ) -> Result<Self, PeetComputeError> {
+        let cipher_suites: Vec<CipherSuite> = client_hello.cipher_suites().collect_vec();
+
+        if cipher_suites.is_empty() {
+            return Err(PeetComputeError::EmptyCipherSuites);
+        }
+
+        let mut supported_tls_versions = Vec::new();
+        let mut supported_protocols = Vec::new();
+        let mut supported_groups = Vec::new();
+        let mut supported_signature_algorithms = Vec::new();
+        let mut psk_key_exchange_mode = None;
+        let mut certificate_compression_algorithms = Vec::new();
+        let mut sorted_extensions = Vec::new();
+
+        for ext in client_hello.extensions() {
+            let id = ext.id();
+            sorted_extensions.push(id);
+
+            if id.is_grease() {
+                continue;
+            }
+
+            match ext {
+                ClientHelloExtension::SupportedVersions(versions) => {
+                    supported_tls_versions.extend(versions.iter().copied());
+                }
+                ClientHelloExtension::ApplicationLayerProtocolNegotiation(alpns) => {
+                    for p in alpns {
+                        match p {
+                            ApplicationProtocol::HTTP_10 => {
+                                supported_protocols.push("1.0");
+                            }
+                            ApplicationProtocol::HTTP_11 => {
+                                supported_protocols.push("1.1");
+                            }
+                            ApplicationProtocol::HTTP_2 => {
+                                supported_protocols.push("2");
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+                ClientHelloExtension::SupportedGroups(groups) => {
+                    supported_groups.extend(groups.iter().copied());
+                }
+                ClientHelloExtension::SignatureAlgorithms(algos) => {
+                    supported_signature_algorithms.extend(algos.iter().copied());
+                }
+                ClientHelloExtension::Opaque { id: ext_id, data }
+                    if *ext_id == ExtensionId::PSK_KEY_EXCHANGE_MODES =>
+                {
+                    if data.len() < 2 {
+                        psk_key_exchange_mode = None;
+                    } else {
+                        psk_key_exchange_mode = Some(data[1]);
+                    }
+                }
+                ClientHelloExtension::CertificateCompression(algs) => {
+                    certificate_compression_algorithms.extend(algs.iter().cloned());
+                }
+                _ => {}
+            }
+        }
+
+        // Sorting integer IDs in lexicographical order
+        sorted_extensions.sort_by_key(|&id| {
+            if id.is_grease() {
```

Comment:
I understand why you are doing it as my approach would indeed not work... However instead of these heap allocations we can use the assumption that it are all u16 numbers to use an array instead, no?

Something like

```
let mut digits = [0; 5];
    for i in (0..5).rev() {
        digits[i] = (n % 10) as u8;
        n /= 10;
    }
```
(again not tested, but hopefully gets the idea across)